core/storage/app_feedback_report/gae_models.py

```diff
@@ -0,0 +1,591 @@
+# Copyright 2021 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Models for Oppia feedback reports from both Android and web."""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+import datetime
+
+from core.platform import models
+import feconf
+import python_utils
+import utils
+
+(base_models,) = models.Registry.import_models([models.NAMES.base_model])
+
+datastore_services = models.Registry.import_datastore_services()
+transaction_services = models.Registry.import_transaction_services()
+
+PLATFORM_CHOICE_ANDROID = 'android'
+PLATFORM_CHOICE_WEB = 'web'
+PLATFORM_CHOICES = [PLATFORM_CHOICE_ANDROID, PLATFORM_CHOICE_WEB]
+
+REPORT_INFO_TO_REDACT = (
+    'user_feedback_other_text_input', 'event_logs', 'logcat_logs')
+
+
+class AppFeedbackReportModel(base_models.BaseModel):
+    """Model for storing feedback reports sent from learners.
+
+    Instances of this model contain information about learner's device and Oppia
+    app settings, as well as information provided by the user in the feedback
+    report.
+
+    The id of each model instance is determined by concatenating the platform,
+    the timestamp of the reports (in ms since epoch, in UTC), and a hash of a
+    string representation of a random int.
+    """
+
+    # We use the model id as a key in the Takeout dict.
+    ID_IS_USED_AS_TAKEOUT_KEY = True
+
+    # The platform (web or Android) that the report is sent from and that the
+    # feedback corresponds to.
+    platform = datastore_services.StringProperty(
+        required=True, indexed=True, choices=PLATFORM_CHOICES)
+    # The ID of the user that scrubbed this report, if it has been scrubbed.
+    scrubbed_by = datastore_services.StringProperty(
+        required=False, indexed=True)
+    # Unique ID for the ticket this report is assigned to (see
+    # AppFeedbackReportTicketModel for how this is constructed). This defaults
+    # to None since initially, new reports received will not be assigned to a
+    # ticket.
+    ticket_id = datastore_services.StringProperty(default=None, indexed=True)
+    # Datetime in UTC of when the report was submitted by the user on their
```

Comment:
Just to check: will the new endpoint controllers require that the timestamps being sent from the Android client be in UTC or will the web server need to make the time conversion? Note that this may be tricky to do reliably on Android since all of the system clock methods can be affected by the user changing their app's time settings. We may need to also store when the report was received by the web server & use that as our primary time reference, and just display the creation date as informational since it can be really wrong.

Reply:
Hmm, I just realized that created_on serves the purpose of when the report was received. Maybe we should be using that instead of this timestamp in cases when time matters. WDYT?

Reply:
Yes for scrubbing / report lifespan it's using the `created_on` timestamp but the report submission date is informational for maintainers. I didn't know that about timestamps on Android, though, is it feasible to convert to UTC once received on the web server?

Reply:
Ideally we can get the submission timestamp in UTC so that it can be standardized for the stats models as well. Based on [this post](https://stackoverflow.com/questions/2818086/android-get-current-utc-time) I think we can get UTC on Android if we use the milliseconds?

Reply:
Ah interesting I never considered that. Your recommendation sounds reasonable -- I'll collect the local timestamp still with the time zone and use the created_on datetime for any data processing. That said, is it still reasonable to use the submitted_on time for the stats model or would you also recommend using the created_on datetime?

It seems inaccurate to use the created_on datetime but I'm not sure how different the submitted_on datetime might be from the actual datetime

Reply:
Hmm, I'm torn. We could change the stats to be based on inflow rather than creation date, but the latter is actually more useful for correlating with events like outages. I suspect the number of cases when submitted_on is wrong isn't often, so we could add it for stats. If we ever want to change this or aggregate for both, we have both timestamps to utilize in the future.

What do you think?

Reply:
I agree with this as well, I was unsure of how likely the device's clock would differ from the true timestamp but I'd also presume that it's unlikely to differ more than a few minutes. I'll similarly use a buffer for stats date validation (about 1 week to be extra safe) -- in any case if we find that the timestamp is a week in the future then it's also less useful since it would correlate less to any important events.